consensus/src/checkpoint.rs

```diff
@@ -112,47 +117,221 @@ mod tests {
         );
     }
 
+    fn gen_num_blocks_in_epochs(slot_fill_rate: f64, slot_fill_rate_delta: f64, n: f64) -> i32 {
+        let protocol_constants = ProtocolConstantsTemp::default();
+        let nums = gen_num_blocks_in_slots(
+            slot_fill_rate,
+            slot_fill_rate_delta,
+            n * protocol_constants.slots_per_epoch.0 as f64,
+        );
+        nums
+    }
+
+    fn gen_spot_root_epoch_position(slot_fill_rate: f64, slot_fill_rate_delta: f64) -> (u32, u32) {
+        //  We need to simulate both the staking epoch and the next staking epoch,
+        //  the root epoch is the staking epoch. The root epoch position this function generates
+        //   is the epoch number of the staking epoch and the block height the
+        //  staking epoch starts at (the simulation of all blocks preceeding the
+        //  staking epoch
+        let root_epoch_int = thread_rng().gen_range(0..100);
+        let root_block_height =
+            gen_num_blocks_in_epochs(slot_fill_rate, slot_fill_rate_delta, root_epoch_int as f64);
+        return (root_epoch_int, root_block_height as u32);
+    }
+
+    fn convert(x: f64) -> i32 {
+        x.round().rem_euclid(2f64.powi(32)) as u32 as i32
+    }
+
+    fn gen_num_blocks_in_slots(slot_fill_rate: f64, slot_fill_rate_delta: f64, n: f64) -> i32 {
+        let min_blocks = n * f64::max(slot_fill_rate - slot_fill_rate_delta, 0.0);
+        let max_blocks = n * f64::min(slot_fill_rate + slot_fill_rate_delta, 1.0);
+        let num_blocks_in_slots = thread_rng().gen_range(min_blocks..max_blocks);
+        let num = convert(num_blocks_in_slots);
+        return num;
+    }
+
+    fn gen_spot(block: &mut ProtocolState) {
```

Comment:
feels like this is missing steps as the one in ocaml: https://github.com/MinaProtocol/mina/blob/af76cb7980d5e81e704120290a850ea9c6f8522e/src/lib/consensus/proof_of_stake.ml#L4250 
?

Reply:
As an added note: we want to be writing the test cases in our own way and not have to follow the code structure as in ocaml impl. Currently the test cases feels like we're porting the test cases.
What we can do is first arrive at a generic and code agnostic description of what the test cases are trying to achieve by reading the ocaml impl, and then re-write them here in our own way.

Reply:
Our gen_spot is different in Mina Ocaml. Our gen_spot() is directly new a, b state then perform operations on the new a, b equal state with different conditions such as `gen_spot_pair_common_checkpoints` for `pair_short_aligned`, `pair_short_misaligned` and `pair_long`